components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1441,6 +1442,185 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
+
+  GetSupportsInterface(contract_address, kERC721MetadataInterfaceId,
+                       std::move(internal_callback));
+}
+
+void JsonRpcService::OnGetSupportsInterfaceERC721Metadata(
+    const std::string& contract_address,
+    const std::string& function_signature,
+    const GURL& network_url,
+    GetERC721MetadataCallback callback,
+    bool is_supported,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
+
+  if (!is_supported) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kMethodNotSupported,
+        l10n_util::GetStringUTF8(IDS_WALLET_METHOD_NOT_SUPPORTED_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetERC721TokenUri,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  // Call tokenURI on the ERC721 contract
+  RequestInternal(eth::eth_call("", contract_address, "", "", "",
+                                function_signature, "latest"),
+                  true, network_url, std::move(internal_callback));
+}
+
+void JsonRpcService::OnGetERC721TokenUri(
+    GetERC721MetadataCallback callback,
+    const int status,
+    const std::string& body,
+    const base::flat_map<std::string, std::string>& headers) {
+  if (status < 200 || status > 299) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  // Parse response JSON that wraps the result
+  GURL url;
+  if (!eth::ParseERC721TokenUri(body, &url)) {
+    mojom::ProviderError error;
+    std::string error_message;
+    ParseErrorResult<mojom::ProviderError>(body, &error, &error_message);
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
+
+  // Obtain JSON from the URL depending on the scheme.
+  // IPFS, HTTPS, and data URIs are supported.
+  // IPFS and HTTPS URIs require an additional request to fetch the metadata.
+  std::string metadata_json;
+  std::string scheme = url.scheme();
+  if (scheme != url::kDataScheme && scheme != url::kHttpsScheme &&
+      scheme != ipfs::kIPFSScheme) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kMethodNotSupported,
+        l10n_util::GetStringUTF8(IDS_WALLET_METHOD_NOT_SUPPORTED_ERROR));
+    return;
+  }
+
+  if (scheme == url::kDataScheme) {
+    if (!eth::ParseDataURIAndExtractJSON(url, &metadata_json)) {
+      std::move(callback).Run(
+          "", mojom::ProviderError::kParsingError,
+          l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+      return;
+    }
+
+    // Sanitize JSON
+    data_decoder::JsonSanitizer::Sanitize(
+        std::move(metadata_json),
+        base::BindOnce(&JsonRpcService::OnSanitizeERC721Metadata,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
```

Comment:
I'm not sure this extra step is needed as I think the body passed into `OnGetERC721TokenUri` is already sanitized?

Reply:
Sorted through DM, it has to be decoded first so we need to run sanitizer against the result of decoded data URI.